File: 0_CONTEXT/Computer_Languages/Fortran/active_inference.f90
! Active Inference Implementation in Fortran

program active_inference
    implicit none
    
    integer, parameter :: num_states = 4
    integer, parameter :: num_observations = 3
    integer, parameter :: num_actions = 2
    
    real(8), dimension(num_states) :: beliefs
    real(8), dimension(num_observations, num_states) :: a_matrix
    real(8), dimension(num_states, num_actions, num_states) :: b_matrix
    real(8), dimension(num_observations) :: c_vector
    real(8), dimension(num_states) :: d_vector
    
    integer :: cycle, observation, action
    
    ! Initialize
    call initialize_model(beliefs, a_matrix, b_matrix, c_vector, d_vector)
    
    print *, "Fortran Active Inference Demo"
    print *, "Initial beliefs:", beliefs
    
    ! Run simulation
    do cycle = 1, 5
        observation = mod(cycle-1, 3) + 1
        call update_beliefs(beliefs, a_matrix, observation)
        action = select_action(beliefs, b_matrix)
        print *, "Cycle", cycle, ": Observation=", observation, ", Action=", action
        print *, "Beliefs:", beliefs
    end do

contains

    subroutine initialize_model(b, a, b_mat, c, d)
        real(8), dimension(:), intent(out) :: b, c, d
        real(8), dimension(:,:), intent(out) :: a
        real(8), dimension(:,:,:), intent(out) :: b_mat
        
        ! Initialize beliefs (uniform)
        b = 1.0d0 / real(num_states, 8)
        
        ! Initialize A matrix (observation likelihood)
        a = reshape([0.8d0, 0.1d0, 0.1d0, &
                    0.1d0, 0.8d0, 0.1d0, &
                    0.1d0, 0.1d0, 0.8d0], [num_observations, num_states])
        
        ! Initialize B matrix (transition likelihood) - simplified
        b_mat = 0.0d0
        b_mat(:,1,:) = reshape([0.9d0, 0.1d0, 0.0d0, 0.0d0, &
                               0.1d0, 0.9d0, 0.0d0, 0.0d0], [num_states, num_states])
        
        ! Initialize C and D vectors
        c = [0.0d0, 0.5d0, 0.0d0]
        d = [0.5d0, 0.5d0]
    end subroutine

    subroutine update_beliefs(b, a, observation)
        real(8), dimension(:), intent(inout) :: b
        real(8), dimension(:,:), intent(in) :: a
        integer, intent(in) :: observation
        
        real(8), dimension(num_states) :: posterior
        real(8) :: total
        
        ! Bayesian update: P(s|o) ∝ P(o|s) * P(s)
        posterior = a(observation, :) * b
        total = sum(posterior)
        
        if (total > 0.0d0) then
            b = posterior / total
        end if
    end subroutine

    function select_action(b, b_mat) result(action)
        real(8), dimension(:), intent(in) :: b
        real(8), dimension(:,:,:), intent(in) :: b_mat
        integer :: action
        
        ! Simplified action selection
        action = 1  ! Default action
    end function

end program active_inference

File: 0_CONTEXT/Computer_Languages/Pascal/active_inference.pas
{ Active Inference Implementation in Pascal }

program ActiveInferenceDemo;

const
  NumStates = 4;
  NumObservations = 3;
  NumActions = 2;

type
  TBeliefsArray = array[1..NumStates] of Real;
  TAMatrix = array[1..NumObservations, 1..NumStates] of Real;
  TBMatrix = array[1..NumStates, 1..NumActions] of Real;
  TCVector = array[1..NumObservations] of Real;
  TDVector = array[1..NumStates] of Real;

var
  beliefs: TBeliefsArray;
  aMatrix: TAMatrix;
  bMatrix: TBMatrix;
  cVector: TCVector;
  dVector: TDVector;

procedure InitializeModel;
var
  i, j: Integer;
begin
  { Initialize beliefs (uniform) }
  for i := 1 to NumStates do
    beliefs[i] := 1.0 / NumStates;
  
  { Initialize A matrix (observation likelihood) }
  aMatrix[1,1] := 0.8; aMatrix[1,2] := 0.1; aMatrix[1,3] := 0.1;
  aMatrix[2,1] := 0.1; aMatrix[2,2] := 0.8; aMatrix[2,3] := 0.1;
  aMatrix[3,1] := 0.1; aMatrix[3,2] := 0.1; aMatrix[3,3] := 0.8;
  
  { Initialize B matrix (transition likelihood) }
  bMatrix[1,1] := 0.9; bMatrix[1,2] := 0.1;
  bMatrix[2,1] := 0.1; bMatrix[2,2] := 0.9;
  
  { Initialize C and D vectors }
  cVector[1] := 0.0; cVector[2] := 0.5; cVector[3] := 0.0;
  dVector[1] := 0.5; dVector[2] := 0.5;
end;

procedure UpdateBeliefs(observation: Integer);
var
  posterior: TBeliefsArray;
  total: Real;
  i: Integer;
begin
  total := 0.0;
  
  { Calculate posterior P(s|o) ∝ P(o|s) * P(s) }
  for i := 1 to NumStates do begin
    posterior[i] := aMatrix[observation, i] * beliefs[i];
    total := total + posterior[i];
  end;
  
  { Normalize }
  if total > 0.0 then
    for i := 1 to NumStates do
      beliefs[i] := posterior[i] / total;
end;

function SelectAction: Integer;
begin
  { Simplified action selection }
  SelectAction := 1;  { Default action }
end;

procedure PrintBeliefs;
var
  i: Integer;
begin
  Write('Beliefs: ');
  for i := 1 to NumStates do begin
    Write(beliefs[i]:4:3);
    if i < NumStates then Write(', ');
  end;
  Writeln;
end;

var
  cycle, observation, action: Integer;

begin
  Writeln('Pascal Active Inference Demo');
  
  InitializeModel;
  Write('Initial '); PrintBeliefs;
  
  for cycle := 1 to 5 do begin
    observation := (cycle - 1) mod 3 + 1;
    UpdateBeliefs(observation);
    action := SelectAction;
    Write('Cycle ', cycle, ': Observation=', observation, ', Action=', action, ', ');
    PrintBeliefs;
  end;
end.
